packages/medusa/src/services/cart.ts

```diff
@@ -674,6 +679,154 @@ class CartService extends TransactionBaseService {
+              })
+            )
+
+            const invalidProducts = areValid
+              .map((valid, index) => {
+                return !valid ? { title: items[index].title } : undefined
+              })
+              .filter((v): v is { title: string } => !!v)
+
+            if (invalidProducts.length) {
+              throw new MedusaError(
+                MedusaError.Types.INVALID_DATA,
+                `The products [${invalidProducts
+                  .map((item) => item.title)
+                  .join(
+                    " - "
+                  )}] must belongs to the sales channel on which the cart has been created.`
+              )
+            }
+          }
+        }
+
+        const lineItemServiceTx =
+          this.lineItemService_.withTransaction(transactionManager)
+        const inventoryServiceTx =
+          this.inventoryService_.withTransaction(transactionManager)
+
+        const existingItems = await lineItemServiceTx.list(
+          {
+            cart_id: cart.id,
+            variant_id: In([items.map((item) => item.variant_id)]),
+            should_merge: true,
+          },
+          { select: ["id", "metadata", "quantity"] }
+        )
+
+        const existingItemsVariantMap = new Map()
+        existingItems.forEach((item) => {
+          existingItemsVariantMap.set(item.variant_id, item)
+        })
+
+        const lineItemsToCreate: LineItem[] = []
+        const lineItemsToUpdate: { [id: string]: LineItem }[] = []
+        for (const item of items) {
+          let currentItem: LineItem | undefined
+
+          const existingItem = existingItemsVariantMap.get(item.variant_id)
+          if (item.should_merge) {
+            if (existingItem && isEqual(existingItem.metadata, item.metadata)) {
+              currentItem = existingItem
+            }
+          }
+
+          // If content matches one of the line items currently in the cart we can
+          // simply update the quantity of the existing line item
+          item.quantity = currentItem
+            ? (currentItem.quantity += item.quantity)
+            : item.quantity
+
+          await inventoryServiceTx.confirmInventory(
+            item.variant_id,
+            item.quantity
+          )
+
+          if (currentItem) {
+            lineItemsToUpdate[currentItem.id] = {
+              quantity: item.quantity,
+              has_shipping: false,
+            }
+          } else {
+            // Since the variant is eager loaded, we are removing it before the line item is being created.
+            delete (item as Partial<LineItem>).variant
+            item.has_shipping = false
+            item.cart_id = cart.id
+            lineItemsToCreate.push(item)
+          }
+        }
+
+        // Update all items that needs to be updated
+        if (Object.keys(lineItemsToUpdate).length) {
+          await Promise.all(
+            Object.keys(lineItemsToUpdate).map(async (variantId) => {
```

Comment:
suggestion(non-blocking): Minor simplification
```suggestion
        const itemsToUpdateKeys = Object.keys(lineItemsToUpdate)
        // Update all items that needs to be updated
        if (itemsToUpdateKeys?.length) {
          await Promise.all(
            itemsToUpdateKeys.map(async (variantId) => {
```